src/java/org/apache/cassandra/utils/AbstractIterator.java

```diff
@@ -44,7 +44,16 @@ public boolean hasNext()
         {
             case MUST_FETCH:
                 state = State.FAILED;
-                next = computeNext();
+                try
+                {
+                    next = computeNext();
+                }
+                catch (Throwable t)
+                {
+                    state = State.FAILED;
+                    close();
```

Comment:
I just wrote https://github.com/datastax/cassandra/pull/969 to take care of some of the leaks we saw this week. In writing that, it raises an important detail: a peer iterator can fail, and the consumer of a collection of iterators must close all of the peers, so a self closing iterator doesn't save us much. Also, with this change, I think we'll iteratively call close on each nested `AbstractIterator`, which could have unintended consequences.

Without redesigning other parts of the code base, I don't think we should add this. What do you think?

Reply:
This is not a comprehensive solution, of course and will not replace the need for the proper error handling.
My line of thought was that at least slowing down memory leak will give production more uptime. 
We can drop this idea, of course.